Das U-Boot: Fork U-Boot v2024.10 to support OreSat am335x custom boards.

chroco opened the pull request:
Edit arch/arm/mach-omap2/Kconfig to include board/psas/oresat board files. Based on the am335x_evm section of this file.

Copy board/ti/ to board/psas/. This includes the common directory and is the template for the oresat family of boards.

Edit board/psas/oresat/Kconfig to use the correct files and to source board/psas/common/Kconfig during build.

Modify board.c board.h and mux.c in board/psas/oresat to boot oresat specific boards from the board and version strings found in the EEPROM. This commit also comments out the replaced code and preserves it for future reference.

Add configs/am335x_oresat_defconfig. This is the base configuraton file u-boot uses to build images. CONFIG_OF_LIST is commented out since it is not used in the current configuration but might be used in the future.

Copy arch/arm/dts/am335x-evm.dts to arch/arm/dts/am335x-oresat.dts and custommize the file. This is the default .dts u-boot uses on startup. Set CONFIG_DEFAULT_DEVICE_TREE="am335x-oresat" inam335x_oresat_defconfig. This .dtb is later replaced from the linux file system in /boot/dtbs/.

Copy include/configs/am335x_evm.h to include/configs/am335x_oresat.h and edit the script to determine which .dtb to load. This script is embedded in a macro as a string and passed the to the u-boot environment.

Add support for oresat-bbb-dev.dtb on the beaglebone black. The .dtb is included in /boot/dtbs during the image builder process. Selection takes place in include/configs/am335x_oresat.h. This .dtb was added for testing purposes and can be used instead of am335x_boneblack.dtb on the beaglebone black by modifying the board and version in the EEPROM.

ThirteenFish commented:
Two initial comments:
- What is gained from forking the beaglebone u-boot fork instead of u-boot directly? Are there specific patches of interest? If yes they should be noted in the initial commit message, if no then the middleman can be skipped and the fork can be from [u-boot](https://github.com/u-boot/u-boot).
- When carrying out of tree patches the commit history itself starts to be important so that it's easy to summarize and see the rationale for the changes. When I rebase the patches on a newer version of u-boot I want to know that I'm not missing anything important. The commits themselves should tell a coherent story of what has changed and the commit message bodies should tell me why and what to pay attention to or be careful with. To that end I don't need to know that you've saved broken work or that you fixed compile errors but I would like to know which files you've used as a template or start of work, and what changes you've made to that base.  

I'd like to see the [history rewritten](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) in to a series of two or threeish commits that's something like
- Just a copy/paste to a new directory of the board files you're starting from, with a commit message that describes why these files are a starting point
- The oresat specific changes: name changes, config options, code that looks for specific boards, that kind of thing. The commit message should note why we need these changes and anything that you had to be especially careful with (if applicable).
- Possibly a third that just adds oresat-bbb-dev? A possible way of framing this is that commit two took things up to parity with what existed before and this patch adds a new feature. Alternatively this could just be in the previous patch to make it easier.

chroco commented:
Thanks for the review Theo!

- I didn't even consider the differences between the beagleboard version and main version of u-boot. Oresat-linux was using a patch of the beagleboard version so I just went with it. I will change this.
- Maybe I'm missing something with my work flow regarding commit history. I commit frequently so I don't lose work and so I can go back to a good state. If I'm experimenting or if I just want to save what I'm working on for later, I like being able to remember what I was doing. This is why there are "save broken work" or "fix compile errors" commits. Is there a better way of doing this? I agree these type of commits aren't helpful to the overall history but they were part of the development process. 

I'll rewrite the history in the way you suggested. Is this a common practice in industry or are people just more mindful with their commit history? And if this is the case, where do they save their intermediate work?

- oresat-bbb-dev.dts is just a renamed am335x-boneblack.dts that I added for testing purposes. It can stay or go

Since I need to redo some of this work, I have a couple of options. I copied the TI directory and then made the necessary changes. I also removed a lot of code that was specific to boards that we don't use. Should I just add our stuff to what's already there or keep the stuff we don't use stripped out of it? I also removed the code that supports uboot having a copy of the DTBs so the option is available to use them. I removed them because they already exist in the image created by image builder.

Maybe these are questions that I should have asked sooner but I feel like PSAS is kind of the wild west as far as development goes and I didn't even know to ask those questions until after all the experiments and testing.

ThirteenFish commented:
> Maybe I'm missing something with my work flow regarding commit history. I commit frequently so I don't lose work and so I can go back to a good state. If I'm experimenting or if I just want to save what I'm working on for later, I like being able to remember what I was doing. This is why there are "save broken work" or "fix compile errors" commits. Is there a better way of doing this? I agree these type of commits aren't helpful to the overall history but they were part of the development process.
> 
> I'll rewrite the history in the way you suggested. Is this a common practice in industry or are people just more mindful with their commit history? And if this is the case, where do they save their intermediate work? 

There's a wide spectrum of people who use git and the skills and processes involved. Typically larger projects or ones with many collaborators will favor more refined histories, it makes review and searching git history easier. In those projects from an individual submitter's standpoint development and submission can be two different stages. The Linux kernel is a standout example, you can for example read their guidelines on [describing](https://kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes) and [separating](https://kernel.org/doc/html/latest/process/submitting-patches.html#separate-your-changes) changes.

Beyond viewing `git log` and pull requests directly there's two other uses for well structured commits that I run into frequently: `git blame` will tell me what commit a particular line came from. I'll use `git blame` to search for additional context as to why a particular line is there or doing what it's doing and if the commit message is descriptive or the commit itself is targeted to a particular feature that helps a lot. The other is `git bisect` where if you have a known good commit and then at a later point in time you know there's a bug, `git bisect` will help you binary search the history for the exact commit that introduced the bug. The process involves checking out the commits one by one and then building and running them so this is much more effective if each commit individually can compile.

On the other end smaller projects or ones with developers who are newer/less familiar with git it doesn't matter as much. PSAS typically has not really cared (though it is appreciated), we're more focused on getting people to use git at all in the first place. Getting people up and running on git at all is a high enough barrier, we wouldn't want to make it any higher.

The reason I'm making an exception to that here in u-boot, and asking for a bit more planned commit history, is that since this will be an out of tree patch series the patches themselves are what we'll be primarily interacting with instead of the broader codebase. We're not maintaining u-boot as a whole, just the handful of bits we want to change. The commits themselves then become the prime location for documentation specific to our changes.

> Where do they save their intermediate work?

A branch with intermediate and exploratory work is a valid and common way of doing this, that then gets cleaned up or recreated afterwards. Also in particular I think you may be able to make good use of `git commit --amend` on the temporary saved work kinds of commits.

> Is there a better way of doing this?

There's a variety of tools in git that you may find useful:
- [The staging area](https://git-scm.com/docs/git-add). When you `git add` a file this records it in a "proto-commit" kind of area where you can prepare more things before doing a `git commit`.  You can either continue editing the file in the working directory and use the staging area as a checkpoint (for example `git restore` will take the file back to the point it was `add`ed to the staging area) or you can further prepare it for commit.
- [`git commit --patch`](https://git-scm.com/docs/git-add#Documentation/git-add.txt---patch) If you don't want for example everything in the staging area to go into a commit you can use the `--patch` or `-p` flag to `git commit` to select only certain hunks. Related, I find `git commit --verbose` or `-v` very handy when creating the commit itself.
- [The stash](https://git-scm.com/docs/git-stash). If you've made changes to the working directory but not enough to be a full commit, and then you need to save them temporarily somewhere out of the way (for example while switching branches or doing a build from a clean repo) `git stash` is handy. It'll put the local changes in a temporary holding area that you can later `git stash pop` when you want to get back to those local changes. The downside is it can be easy to forget that there are stashed changes, I've had repos where I look at the stash months later realizing I totally forgot about some set of stashed changes.
- [`git commit --amend`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---amend). If you've made a commit and then realized you need to add to it, maybe correct formatting or fix a compile error, then you can `git add` the changes and `git commit --amend` to rewrite the previous commit with the new changes included. Be aware that this does rewrite history so if you've already pushed the commit  somewhere you'll have to force-push it again. If other people have already cloned the un-amended commit then resolving their copies can start to get a bit messy. Try to keep history rewrites done before you push, although it's not the end of the world if you force-push rewritten history plausibly before anyone else has pulled.
- [`git commit --fixup`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt). If you want to amend a commit that isn't the most recent one you can use a fixup commit, which will create a new commit referencing the older one. Before submitting upstream you can then [`git rebase --autosquash`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---autosquash), and it will rewrite history for you, applying all the `fixup` (or `amend`, `squash`) patches to their marked targets.
- [`git rebase --interactive`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---interactive). This is sort of like the full power history editor, there's a lot you can do from rearranging to splitting and merging commits and other documentation is probably more comprehensive than what I can describe here. One popular use is squashing where a bunch of commits are collapsed into one.

Feel free to mix and match tools and techniques, and with a bit of practice you can also start to anticipate which sets of changes would make for good individual commits. [Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) has more discussion and also there's a number of other blogs out there that describe the rebase/squash/fixup workflow.

> oresat-bbb-dev.dts is just a renamed am335x-boneblack.dts that I added for testing purposes. It can stay or go

Seems fine to keep to me, it's easier to acquire an off the shelf part than a custom oresat board if other people also want to hack on this. 

> Since I need to redo some of this work, I have a couple of options. I copied the TI directory and then made the necessary changes. I also removed a lot of code that was specific to boards that we don't use. Should I just add our stuff to what's already there or keep the stuff we don't use stripped out of it?

Stripping out unused parts seems nice. Having a commit that explicitly contains all the removed code could make it easier to understand later on, otherwise I'd have to find the original file and diff it with an external tool which would be harder. It's also easy to miss that code was removed otherwise too.

> I also removed the code that supports uboot having a copy of the DTBs so the option is available to use them. I removed them because they already exist in the image created by image builder.

This is probably a good idea too but I've not reviewed it in depth so I can't really make a firm statement? I'll be able to make a firmer statement after some reading on my part. This is probably also another good candidate for a standalone commit.

chroco committed:
Das U-Boot: Fork U-Boot v2024.10 to support OreSat am335x custom boards.

Edit arch/arm/mach-omap2/Kconfig to include board/psas/oresat board
files. Based on the am335x_evm section of this file.

Copy board/ti/ to board/psas/. This includes the common directory and
is the template for the oresat family of boards.

Edit board/psas/oresat/Kconfig to use the correct files and to source
board/psas/common/Kconfig during build.

Modify board.c board.h and mux.c in board/psas/oresat to boot oresat
specific boards from the board and version strings found in the
EEPROM. This commit also comments out the replaced code and preserves
it for future reference.

Add configs/am335x_oresat_defconfig. This is the base configuraton
file u-boot uses to build images. CONFIG_OF_LIST is commented out
since it is not used in the current configuration but might be used
in the future.

Copy arch/arm/dts/am335x-evm.dts to arch/arm/dts/am335x-oresat.dts and
custommize the file. This is the default .dts u-boot uses on startup.
Set CONFIG_DEFAULT_DEVICE_TREE="am335x-oresat" inam335x_oresat_defconfig.
This .dtb is later replaced from the linux file system in /boot/dtbs/.

Copy include/configs/am335x_evm.h to include/configs/am335x_oresat.h
and edit the script to determine which .dtb to load. This script is
embedded in a macro as a string and passed the to the u-boot environment.

Add support for oresat-bbb-dev.dtb on the beaglebone black. The .dtb
is included in /boot/dtbs during the image builder process. Selection
takes place in include/configs/am335x_oresat.h. This .dtb was added
for testing purposes and can be used instead of am335x_boneblack.dtb
on the beaglebone black by modifying the board and version in the EEPROM.

ThirteenFish reviewed (changes requested):
I'm still working through the technical parts, but in the meantime I have some requests on style:
 - Prefer deleting code instead of commenting it out. This shows up in diffs and there's no need to keep around dead code. It will always be preserved in git history.
 - Try to mimic the existing code style.
   - It uses a single tab instead of two spaces for indent
   - `if` statements have the braces at the end of the line instead of the following line: `if (foo) {`
   - Single line `if`s omit braces
   - Trim whitespace at the trailing ends of lines, either spaces after the end of text or blank lines with only spaces/tabs on them
   - Leave comment style as is, they typically use block `/**/` over line `//`
 - `CONFIG_XPL_BUILD` is undefined. This symbol appears to have been introduced in a patch on October 11th but this repo is forked off of the release on October 7. I think it's good to base off of a stable release so we'll probably want to update this patch. When v2025.01 comes out we'll have to remember to convert back from `SPL` to `XPL`.

ThirteenFish commented on board/psas/oresat/board.c:
This change doesn't reflect the intended behavior of the function. If I understand correctly this function answers the question "Does the current board I'm running on require the device tree named in `name`". That's why all the `strcmp()` are there. This picks the first device tree and says yes, this one is mine.

ThirteenFish commented on configs/am335x_oresat_defconfig:
`defconfigs` are like the default starting point configs for individual projects to modify. Since we'll probably want to do a bunch of modifications ourselves the config should probably live in `oresat-linux` (and not be named `_defconfig`). This will also help with change visibility, we'll want to separate out our specific config from u-boot patches. Also we'll not really want to update the submodule every time we make a config change. You can pass an arbitrary config to a Kbuild system through the environment variable `KCONFIG_CONFIG` so for example `KCONFIG_CONFIG=am335x_oresat_config make -C u-boot`.
